Blog post about RANLUX++ (#465)
root-project/web: blog/_posts/2020-11-24-ranluxpp.md

---
title: "Implementing and tuning RANLUX++ for ROOT"
layout: archive
author: Jonas Hahnfeld
---

ROOT includes different implementations of pseudorandom number generators (PRNGs):
 * [`TRandom`][TRandom] is a Linear Congruential Generator (LCG) based on the BSD `rand` generator.
 * [`TRandom1`][TRandom1] implements the RANLUX generator described by Fred James in 1990.
 * [`TRandom3`][TRandom3] is based on the Mersenne Twister generator.
   By default, this implementation is used for `gRandom`.
 * `TRandomMixMax` is an instantiation of the parametric [`MixMaxEngine`][MixMaxEngine].

[TRandom]: https://root.cern/doc/master/classTRandom.html
[TRandom1]: https://root.cern/doc/master/classTRandom1.html
[Trandom3]: https://root.cern/doc/master/classTRandom3.html
[MixMaxEngine]: https://root.cern/doc/master/classROOT_1_1Math_1_1MixMaxEngine.html

> This post discusses the implementation of the recently added `TRandomRanluxpp`.

Reviewer: In order to have a nicer "preview" on the front page (see https://root.cern/465 ) I would suggest to expand this last sentence a bit and put it first (for example, just to give you an idea: "ROOT offers many different random number generators, and we just added one that we think you might be interested in! In this post I will go over what RANLUX++ is and how I recently implemented it and optimized it for ROOT.").

The list of available PRNGs could then be moved under `## Why so many PRNGs?`, or just under this first introductory paragraph.

Author: Adapted, let me know what you think!

Reviewer: perfect!

## Why so many PRNGs?

There exist many properties to evaluate PRNGs, with the easiest being its period:

> It captures how many random numbers can be generated before the sequence "wraps around" and repeats.

Reviewer: ```suggestion
it captures how many random numbers can be generated before the sequence "wraps around" and repeats.
```

For obvious reasons, the period of a PRNG cannot be longer than its state.
This is one of the problems of the oldest LCGs with a state of 32 bits or less.

However, there are many more statistical properties that are at least equally important.
A widely used collection of empirical tests is [TestU01][TestU01] by L'Ecuyer and Simard.
It has to be noted that the tests can only prove the existence of defects.
That is, passing all tests does not mean that a PRNG is free of errors.

[TestU01]: http://simul.iro.umontreal.ca/testu01/tu01.html

As with many other design issues, PRNGs need to make a certain trade-off:

> Fast generators with a small state will usually fail many statistical tests.

Reviewer: ```suggestion
fast generators with a small state will usually fail many statistical tests.
```

On the other hand, good generators are typically slower or use a much larger state.
In fact, most generators will fail one test or another, including the widely used Mersenne Twister.
Notable exceptions are RANLUX and MIXMAX that pass all currently known tests.
However, they are rather slow (RANLUX) or require a large state (MIXMAX).

## RANLUX++

[RANLUX++][paper] is an LCG implementation derived from the ideas of RANLUX.
It aims to provide better performance while maintaining the good properties of RANLUX.
The generator uses a single 576 bit number as state.

> Additionally it needs to remember from which offset to take the next random bits.

Reviewer: I don't understand this (because I'm not expert enough on PRNGs, for sure, but consider I'm the average ROOT blog reader)

Author: Hopefully clarified a bit

Reviewer: :+1:

As an LCG, it performs multiplications and modulo operations to advance the state.

[paper]: https://arxiv.org/pdf/1705.03123.pdf

Arithmetic on 576 bit numbers is way beyond what current hardware offers.
Instead the state is portably stored as 9 numbers of 64 bits each.
On this representation, multiplication and the modulo operation must thus be done in software.
For the multiplication, this is very similar to the way multiplications taught in high school.
The modulo operation can take advantage of the known modulus (for details, refer to section 2.1 of the [paper][paper]).

For efficiency reasons, Sibidanov [implemented the routines in x86 assembly][ranluxpp].

> This results in impressive performance, but is not portable across architecture supported by ROOT.

Reviewer: ```suggestion
This results in impressive performance, but it is not portable across all architectures supported by ROOT.
```

So how much slower would a portable implementation be?
After implementing RANLUX++ for ROOT, it turns out: not much.

[ranluxpp]: https://github.com/sibidanov/ranluxpp/

> The [initial version][initial] added to ROOT takes around 30ns per random `double`.

Reviewer: ```suggestion
The [initial version][initial] added to ROOT takes around 30ns per random `double` sampled.
```

For comparison, the implementation by Sibidanov comes in at around 8ns per number.
This was measured with 10 repetitions of a [microbenchmark][benchmark] with a standard deviation of less than 1%.
The numbers are from a single core of an AMD Ryzen 9 3900 running CentOS 8.2.2004.
The default compiler is GCC in version 8.3.1, but I will present more numbers in the following.

[initial]: https://github.com/root-project/root/commit/36cc2a820a2b5a687bb603fcaac5f2c09e574ea2
[benchmark]: https://github.com/root-project/rootbench/pull/200

## Performance tuning

30ns per number is already quite impressive and twice as fast compared to RANLUX as implemented in `TRandom1` (around 62ns per number).
But can we do even better? Yes!

> The graph below (created with ROOT) shows the evolution of performance after tuning.

Reviewer: ```suggestion
The graph below shows the evolution of performance after tuning.
```

what else? :stuck_out_tongue:

In addition to GCC 8.3.1, I also include numbers for Clang 9.0.1 and GCC 7.5.0.
While the former two are packaged for CentOS 8, I compiled GCC 7.5.0 from the official sources.
The dashed horizontal line represents the assembly version by Sibidanov.
Details about the individual changes are described in the following sections.

![Performance of RANLUX++]({{ '/assets/images/ranluxpp.png' | relative_url }})

### Loop unrolling

The portable code relies on loops to implement the multiplication and modulo operation in software.
This induces overhead when evaluating the exit condition and incrementing the loop variable.
Furthermore, the loop structure leads to jumps in the code that hinder instruction level parallelism.
Fortunately loop unrolling is a well known compiler optimization to tackle this problem:

> It duplicates the loop body to enable other optimization passes and generate more efficient code.

Reviewer: ```suggestion
it duplicates the loop body to enable other optimization passes and generate more efficient code.
```

In case of constant loop bounds, it is even possible to fully unroll the loop.

One disadvantage of loop unrolling is the increased code size.
For this reason, compilers are conservative in their unrolling heuristics.
During experiments, I found that unrolling of certain loops consistently improves performance.
[By adding `#pragma` directives][unrolling], it is possible to help the compiler.
As can be seen in above graph, this improves performance for GCC 8.3.1 and Clang 9.0.1 by up to 30%.
GCC 7.5.0 does not benefit from the change because `#pragma GCC unroll` was only added in GCC 8.

[unrolling]: https://github.com/root-project/root/commit/0f0a8077a9e9902e08ff0f8cc93644f00ef33559

### Overflow handling

For the multiplication of two 576 bit numbers, the portable code is constrained by the available data types:

> It can at most multiply two 32 bit values at a time to stay within the 64 bits of `uint64_t`.

Reviewer: ```suggestion
it can at most multiply two 32 bit values at a time to stay within the 64 bits of `uint64_t`.
```

### Carry propagation

Additional tests showed inferior performance with older compilers.
Part of this is related to missing loop unrolling because the `#pragma` is not available (see above).
An additional problem is the generation of jumping code to propagate carry bits in case of overflows.
Instead of relying on inline assembly code, I found a [portable solution][carry] after [refactoring the common code][prep].
The modification results in good code for all tested compilers.
For the older GCC 7.5.0 in particular, it improves performance by more than 20%.

[prep]: https://github.com/root-project/root/commit/f081d391e6b05d85520701b940c008107006313c
[carry]: https://github.com/root-project/root/commit/28dd98b3f7d4a3a43a331b9f6483fcdc19244791

## Conclusion

In this blog post, I discussed the implementation of RANLUX++ and its performance.
After tuning, the average time per number is down to around 9ns, very close to the assembly version.
It is slightly slower than `TRandom3` currently used for `gRandom` (around 3ns per number).
However, RANLUX++ inherits mathematically proven properties from RANLUX and passes all tests in TestU01.

> For this reason, it might replace the default `gRandom` in the future.

Reviewer: please add a mention of the ROOT version users will need to get the goodies!

---
title: "Implementing and tuning RANLUX++ for ROOT"
layout: archive
author: Jonas Hahnfeld
---

ROOT comes with support for different pseudorandom number generators (PRNGs).
This post discusses the recent implementation of RANLUX++ and how I tuned its performance.

The added `TRandomRanluxpp` joins the group of available RNGs in ROOT:

 * [`TRandom`][TRandom] is a Linear Congruential Generator (LCG) based on the BSD `rand` generator.
 * [`TRandom1`][TRandom1] implements the RANLUX generator described by Fred James in 1990.
 * [`TRandom3`][TRandom3] is based on the Mersenne Twister generator.
   By default, this implementation is used for `gRandom`.
 * `TRandomMixMax` is an instantiation of the parametric [`MixMaxEngine`][MixMaxEngine].

There are a few more variants that are linked [in the documentation][TRandom].

> [TRandom]: https://root.cern/doc/master/classTRandom.html

Author: I'm not sure I get this comment: Do you mean we should remove the entire paragraph? The intent is to show that RANLUX++ is certainly not the first PRNG, but then go on to explain why it still makes sense to add another one.

> [TRandom]: https://root.cern/doc/master/classTRandom.html

Reviewer 2: I was referring to the second variants after the bullet points. I would just leave the line:
"There are a few more variants that are linked [in the documentation][TRandom]."

> [TRandom]: https://root.cern/doc/master/classTRandom.html

Author: Ah, you mean the list of links starting with `[TRandom]`? That is just a Markdown way of putting links after the paragraph where they are used to keep the text readable. The following:
```
[link text][label]

[label]: https://example.com/bla
```
will render as a single "link text" that refers to https://example.com/bla. See https://root.cern/465/blog/ranluxpp/ for how this looks like after building the site.

## Why so many PRNGs?

There exist many properties to evaluate PRNGs, with the easiest being its period:
it captures how many random numbers can be generated before the sequence "wraps around" and repeats.
For obvious reasons, the period of a PRNG cannot be longer than its state.
This is one of the problems of the oldest LCGs with a state of 32 bits or less.

However, there are many more statistical properties that are at least equally important.
A widely used collection of empirical tests is [TestU01][TestU01] by L'Ecuyer and Simard.
It has to be noted that the tests can only prove the existence of defects.
That is, passing all tests does not mean that a PRNG is free of errors.

[TestU01]: http://simul.iro.umontreal.ca/testu01/tu01.html

As with many other design issues, PRNGs need to make a certain trade-off:
fast generators with a small state will usually fail many statistical tests.
On the other hand, good generators are typically slower or use a much larger state.
In fact, most generators will fail one test or another, including the widely used Mersenne Twister.
Notable exceptions are RANLUX and MIXMAX that pass all currently known tests.

> However, they are rather slow (RANLUX) or require a large state (MIXMAX).

Author: I think it's important to capture that different PRNGs make trade-offs between speed, state, and quality. I will add the note about mixing below.

> However, they are rather slow (RANLUX) or require a large state (MIXMAX).

Reviewer 2: I agree, but the problem is that something is quite controversial. For example, there is MRNG32ka which passes tests and having a small state, and also MIXMAX17 that is fast and passes tests and the state is not so large. 
Maybe just remove:
"However, they are rather slow (RANLUX) or require a large state (MIXMAX)."

> However, they are rather slow (RANLUX) or require a large state (MIXMAX).

Author: Sounds good.

## RANLUX++

[RANLUX++][paper] is an LCG implementation derived from the ideas of RANLUX.
It aims to provide better performance while maintaining the good properties of RANLUX.
The generator uses a single 576 bit number as state.

> When invoked, it hands out the next unused bits that the generator remembers as on offset.

Reviewer: ```suggestion
When invoked, it hands out the next unused bits that the generator remembers as an offset.
```

Once all bits are exhausted, the LCG performs a multiplication and a modulo operation to advance the state.

> [paper]: https://arxiv.org/pdf/1705.03123.pdf

Reviewer 2: I would add here a sentence mentioning that Ranlux++ allows to use a much higher luxury level (i.e. numbers which  needs to be skipped to obtain a chaotic behavior)  for free. For example the current implementation use a p=2048, much higher that the default of Ranlux (p=389).

Author: I intentionally avoided talking about luxury levels because you can only understand this after reading the referenced papers in depth. I wouldn't even claim that I fully get the relationship to the implementation of RANLUX++: the connection was obvious with RANLUX where skipping more states meant better quality, but in the LCG formulation it's just an exponentiation...

Reviewer 2: I think some mention should be kept, because this is in my opinion a strength of this generator. Maybe we should not go into much detail and refer to the paper

Author: Sure, I added notes both here and in the conclusion that RANLUX++ uses more bits of randomness and has better properties "for free" (in terms of performance) that are understood in theory.

Arithmetic on 576 bit numbers is way beyond what current hardware offers.
Instead the state is portably stored as 9 numbers of 64 bits each.
On this representation, multiplication and the modulo operation must thus be done in software.

> For the multiplication, this is very similar to the way multiplications taught in high school.

Reviewer: ```suggestion
For the multiplication, this is very similar to the way multiplications taught in high school work.
```

> For the multiplication, this is very similar to the way multiplications taught in high school.

Reviewer: ```suggestion
For the multiplication, this is very similar to the way multiplications are taught in high school.
```

The modulo operation can take advantage of the known modulus (for details, refer to section 2.1 of the [paper][paper]).

For efficiency reasons, Sibidanov [implemented the routines in x86 assembly][ranluxpp].
This results in impressive performance, but it is not portable across all architectures supported by ROOT.
So how much slower would a portable implementation be?

> After implementing RANLUX++ for ROOT, it turns out: not much.

Reviewer: ```suggestion
After implementing RANLUX++ for ROOT, it turns out: not much!
```

[ranluxpp]: https://github.com/sibidanov/ranluxpp/

The [initial version][initial] added to ROOT takes around 30ns per random `double` sampled.
For comparison, the implementation by Sibidanov comes in at around 8ns per number.
This was measured with 10 repetitions of a [microbenchmark][benchmark] with a standard deviation of less than 1%.
The numbers are from a single core of an AMD Ryzen 9 3900 running CentOS 8.2.2004.
The default compiler is GCC in version 8.3.1, but I will present more numbers in the following.

[initial]: https://github.com/root-project/root/commit/36cc2a820a2b5a687bb603fcaac5f2c09e574ea2
[benchmark]: https://github.com/root-project/rootbench/pull/200

## Performance tuning

> 30ns per number is already quite impressive and twice as fast compared to RANLUX as implemented in `TRandom1` (around 62ns per number).

Reviewer 2: I would mention here that a new C Ranlux implementation by M. Luscher (see http://luscher.web.cern.ch/luscher/ranlux/) is also around  30 ns/call (see https://link.springer.com/article/10.1007/s41781-019-0034-3)

The [change taking advantage of `__int128`][native] improves performance of all three compilers by more than 40%!

[native]: https://github.com/root-project/root/commit/e37bfe378c50e5d1f638a44801843123c678ce19

### Carry propagation

Additional tests showed inferior performance with older compilers.
Part of this is related to missing loop unrolling because the `#pragma` is not available (see above).
An additional problem is the generation of jumping code to propagate carry bits in case of overflows.
Instead of relying on inline assembly code, I found a [portable solution][carry] after [refactoring the common code][prep].
The modification results in good code for all tested compilers.
For the older GCC 7.5.0 in particular, it improves performance by more than 20%.

[prep]: https://github.com/root-project/root/commit/f081d391e6b05d85520701b940c008107006313c
[carry]: https://github.com/root-project/root/commit/28dd98b3f7d4a3a43a331b9f6483fcdc19244791

## Conclusion

In this blog post, I discussed the implementation of RANLUX++ and its performance.
After tuning, the average time per number is down to around 9ns, very close to the assembly version.

> It is slightly slower than `TRandom3` currently used for `gRandom` (around 3ns per number).

Reviewer 2: I would also add here that TRandom3 is only a 32 bits generators while new RANLUX++ is a 52 bits

> However, RANLUX++ inherits mathematically proven properties from RANLUX and passes all tests in TestU01.

Reviewer 2: I would stress here the theoretical properties and the higher luxury level than standard RANLUX
